app/src/main/kotlin/com/wire/android/ui/home/messagecomposer/state/MessageComposerState.kt

```diff
@@ -20,403 +20,151 @@
+            messageComposition = messageCompositionHolder.messageComposition,
+            selfDeletionTimer = selfDeletionTimer,
+        )
     }
 
-    fun toEditMessage(messageId: String, originalText: String, originalMentions: List<MessageMention>) {
-        messageComposeInputState = MessageComposeInputState.Active(
-            messageText = TextFieldValue(text = originalText, selection = TextRange(originalText.length)),
-            type = MessageComposeInputType.EditMessage(messageId, originalText)
+    return remember {
+        MessageComposerStateHolder(
+            messageComposerViewState = messageComposerViewState,
+            modalBottomSheetState = modalBottomSheetState,
+            messageCompositionInputStateHolder = messageCompositionInputStateHolder,
+            messageCompositionHolder = messageCompositionHolder,
+            additionalOptionStateHolder = AdditionalOptionStateHolder(),
+            tempWritableImageUri = tempWritableImageUri,
+            tempWritableVideoUri = tempWritableVideoUri,
         )
-        mentions = originalMentions.map { it.toUiMention(originalText) }
-        quotedMessageData = null
-        inputFocusRequester.requestFocus()
     }
+}
 
-    fun showRichTextEditingOptions() = changeRichTextEditingOptionsVisibility(true)
-    fun hideRichTextEditingOptions() = changeRichTextEditingOptionsVisibility(false)
-
-    private fun changeRichTextEditingOptionsVisibility(show: Boolean) {
-        (messageComposeInputState as? MessageComposeInputState.Active)?.let { activeState ->
-            when (val currentType = activeState.type) {
-                is MessageComposeInputType.NewMessage -> {
-                    messageComposeInputState = activeState.copy(
-                        type = currentType.copy(
-                            richTextFormattingOptionsDisplayed = show,
-                            attachmentOptionsDisplayed = false
-                        )
-                    )
-                }
-
-                is MessageComposeInputType.EditMessage -> {
-                    messageComposeInputState = activeState.copy(
-                        type = currentType.copy(
-                            richTextFormattingOptionsDisplayed = show
-                        )
-                    )
-                }
-
-                is MessageComposeInputType.SelfDeletingMessage -> {
-                    messageComposeInputState = activeState.copy(
-                        type = currentType.copy(
-                            richTextFormattingOptionsDisplayed = show,
-                            attachmentOptionsDisplayed = false
-                        )
-                    )
-                }
-            }
-        }
-    }
+/**
+ * Class holding the whole UI state for the MessageComposer, this is the class that is used by the out-side world to give the control
+ * of the state to the parent Composables
+ */
+class MessageComposerStateHolder(
```

Comment:
```suggestion
class MessageComposerViewModel(
```

This looks like a ViewModel for the MessageComposer more than the actual `MessageComposerViewModel` we currently have.

Our current `MessageComposerViewModel` has things like `deleteMessage` and `startSelfDeletingCounter`, which are _not_ related to composing of messages.

This one actually does have everything related to what an actual ViewModel for the message composer would do.

Reply:
I think @trOnk12 is following Google recommendation by making `stateHolders` for UI logic
https://youtu.be/pCX9wvu-Bq0?t=535
But I do agree with about `MessageComposerViewModel` that have things not related to message composing.

Reply:
I definetly do not agree that the ViewModel should care about the state of the input types if the attachment is visible or not

Reply:
I agree that some of the stuff inside MessageComposerStateHOlder are not 100 % related to the UI and could be moved but I would not say that this is a ViewModel

Reply:
the next step wqould be to refactor ConversationScreen and make the MessageComposerViewModel live inside the MEssageComposer, then we could move some stuff to the ViewMOdel from here

Reply:
Sorry. I didn't express myself clearly. I'd like to put emphasis in **_italics_**.

> This looks like a `ViewModel` for the `MessageComposer` **_more than the actual_** `MessageComposerViewModel` we currently have.

So, my point is that the current one has many responsibilities it shouldn't have.

Sure, there's some UI-specific stuff like `inactive()`, `active()` and focus stuff. But the `toReply()` and `toEdit()` alone, which update the composition state, are closer to a proper ViewModel for the MessageComposer than the current one.

-----

I wasn't aware that of this practice of using the suffix "StateHolder" for "UI-specific ViewModels". I'm likely wrong here to think of them as ViewModels, but the name ViewModel is supposed to mean "the model of a view" 🤷🏼‍♂️, and the way I see it being used is pretty similar to:
- Expose state
- Expose functions to update state
- Observe their values and change the UI accordingly

Only difference is that it isn't gonna call business-logic. Which I believe it's a significant distinction and I see now how using the same name can be misleading.

Anyway, today I've learned. If it's a known standard, as I see it is now, I'm happy to know that we're adopting it.

Reply:
> the next step wqould be to refactor ConversationScreen and make the MessageComposerViewModel live inside the MEssageComposer, then we could move some stuff to the ViewMOdel from here

Yes!!!! 10x yes!

I see stuff like `EditMessageMenu` which is misleading and could have a better name too.

There is much room for improvement, and I believe you've already taken most of it out of the way regarding the composition.

Reply:
I feel like since day 1 of wokring on Wire we are constantly loosing the meaing of the ViewModel ... there  were so many discussion on what ViewModel should be doing but I feel like we still have our own definitions of the ViewModel...

Reply:
Iny me eyes and from what I read online and from my experience which is of course not huge this is diffenetly not what ViewModel should be doing, in the old days of XML or CustomView, this would also not live inside the ViewModel,. this is a behaviour baked in into the View not ViewModel